Refactor whisper asr pipeline to include language too.

Narsil opened the pull request:
# What does this PR do?

## Why a refacto of this magnitude ?

- Earlier iterations tried to keep in line with `pipeline`, which started a cascade of `if ... else ` just for `whisper`.
- Those cascade of functions where all loosing the `language` information which would make it tough to include
  `language` in the return.
- Since we have chunking, potentially we have several `language` codes within a single file making this a segmentation problem, not a classification problem. (So using `detect_language(..) -> str` not really applicable to the pipeline).


Given this problem space, it was simpler to reimplement the whole thing within the `tokenizer` akin to `_build_conversational_inputs` which can allow lots of specifities within each model.
Here the new `_decode_asr` doesn't contain any information outside the tokenizer.



## How does it work ?

Hopefully inline comments should be the most comprehensive tools.
The tokenizer, needs `stride`, expressed in `seconds` and `time_precision` to convert from token space to seconds space.
We could do this outside of `tokenizer` if necessary (everything kept in tokenizer space) but since `time_precision` is already used as args of some methods, we can use this, so the output of this function doesn't need to be converted back.

=========== Overview ============
 - iterate over all outputs
 - all tokens within output
 - Each token can be
   - language token
   - special token
   - timestamp token
   - text token
 - We accumulate the text tokens.
 - We split on end timestamps
 - Lots of complexity comes from stride and timestamps

Most importantly we need to handle strides, and timestamps within strides. In order to handle those, we're simply not splitting on them, and handling merges later.

All the merges is relatively simple, we find the maximum overlapping sequence.
Small optimization: The overlap sequence might contain errors/conflicts. We're choosing from the previous sequence on the left side of the overlap, and next sequence on right side of the overlap. Since those tokens should correspond to the same audio, midway should be correct most of the time. 

## Benchmark

Courtesy of @ArthurZucker 
```python
# Whisper HF captioning
from evaluate import load
from datasets import load_dataset
import numpy as np
from transformers import pipeline
import time
import whisper


libri = load_dataset("librispeech_asr", "clean", split="test")


batch = len(libri)
start_dataset = 0

# batch = 10
# start = 0 * batch

models = ["tiny", "tiny.en", "base", "base.en", "small", "small.en", "medium", "medium.en", "large", "large-v2"]
# models = ["tiny.en"]
# models = ["large-v2"]

for model_name in models:
    speech_recognizer = pipeline(
        task="automatic-speech-recognition", model=f"openai/whisper-{model_name}", framework="pt", device=1
    )
    model = whisper.load_model(f"{model_name}" + "-v1" if model_name == "large" else f"{model_name}").to(
        device="cuda:2"
    )
    for offset in range(start_dataset, len(libri), batch):

        iterator = range(offset, offset + batch)
        speech_file = np.concatenate([libri[i]["audio"]["array"] for i in iterator])
        labels = " ".join([libri[i]["text"] for i in iterator])

        start = time.time()
        hf = speech_recognizer(
            [speech_file],
            return_timestamps=True,
            chunk_length_s=30,
            stride_length_s=[4, 4],
            batch_size=32,
            ignore_warning=True,
            num_workers=1,
        )[0]
        end = time.time()
        # print(res)
        print(f"model : {model_name}\nhf time : ", end - start)
        # print(res["text"])
        with open("hf.txt", "w") as f:
            f.write(hf["text"])

        norm_labels = speech_recognizer.tokenizer._normalize(labels)
        norm_res = speech_recognizer.tokenizer._normalize(hf["text"])

        wer = load("wer")
        hf_wer = wer.compute(predictions=[norm_res], references=[norm_labels])
        print("hf wer :", hf_wer)
```

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts and @NielsRogge
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer: @stas00, Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

Narsil committed:
[WIP] whisper refacto to support language output.

Narsil committed:
Handling merges.

Narsil committed:
A bit more cleanup and comments.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_21427). All of your documentation changes will be reflected on that endpoint.

ArthurZucker commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -705,3 +705,278 @@ def get_decoder_prompt_ids(self, task=None, language=None, no_timestamps=True):
+                        pass
+                elif token >= timestamp_begin:
+                    # 3/ Timestamp token
+                    time = (token - timestamp_begin) * time_precision + time_offset
```
The time here is wrong, it only includes the negative sum of the stride left.

ArthurZucker commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -705,3 +705,278 @@ def get_decoder_prompt_ids(self, task=None, language=None, no_timestamps=True):
+                        # There is only a merge to handle when there is more than one
+                        # sequence here. That happens only when the last timestamp
+                        # happened in the stride of the audio
+                        resolved_tokens = _find_longest_common_sequence(previous_tokens)
```
My only issue is that here we are looking at all the previous tokens, this is prone to a lot of error + slow

Narsil replied:
This seems actually somewhat faster than before (not sure because `num_workers=1` in the bench seems to gain so much speedup).

This only takes the last *unused* tokens, so it won't take the entire previous audio for processing.

This function is indeed quite slow, but it's quite necessary to handle all the weird cases of timestamps outputs within whisper (like I've seen pretty much all cases now, no overlap, regular  ([a, b, c], [b, c, d]) left included in right ([b, c], [a, b, c]) right included in left ([a, b, c], [b, c]).

The parts where output is the worst is actually linked to the model where we get something like

([a, b, c, d], [b,c]]) and `d` is actually correct. But we cannot take it in general, because in general either `d` will be contained in the next token, or it is hallucinated and actually incorrect (but there are *some* cases where it's correct and the follow-up misses it)

Narsil committed:
Many improvements.

Lots of details everywhere.

Narsil committed:
Cleanup old code and tests.

Narsil committed:
Handle lone timestamp tokens (just recover when something bad happens).

Narsil committed:
Adding return_language example.

Narsil committed:
No ffmpeg.

Narsil committed:
Hmm.

sgugger commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -705,3 +705,272 @@ def get_decoder_prompt_ids(self, task=None, language=None, no_timestamps=True):
+        handle the various options not allowed in other seq2seq models
+        """
+
+        # =========== Overview ============
+        # - iterate over all outputs
+        # - all tokens within output
+        # - Each token can be
+        #   - language token
+        #   - special token
+        #   - timestamp token
+        #   - text token
+        # - We accumulate the text tokens.
+        # - We split on end timestamps
+        # - Lots of complexity comes from stride and timestamps
```
Might be good to add comments in the code below matching this summary.

Narsil replied:
Some of it can be aligned, except the *complexity* one, the `each token` which unfortunately is a bit spread out (stop conditions depend on many things to be as early as possible, and still handle conflicts)

Th rest should be aligned.

sgugger commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -705,3 +705,272 @@ def get_decoder_prompt_ids(self, task=None, language=None, no_timestamps=True):
+        right_stride_start = None
+
+        all_special_ids = set(self.all_special_ids)
+        all_special_ids.add(50256)
```
Why? A comment would be appreciated.

Narsil replied:
This is because `pad_token_id` is sometimes set as 50256, which is a regular token, not `all_special_ids`.

This leads to weird conflict handling (since those ids get interpreted as real text, hence depend on batch size). 

This is imo an issue with the config of the models. However that was done on purpose iiuc. @ArthurZucker do you remember/have a link with the original issue ?

I removed this because it doesn't belong here regardless.

sgugger commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -705,3 +705,272 @@ def get_decoder_prompt_ids(self, task=None, language=None, no_timestamps=True):
+            right = np.array(right_sequence[right_start:right_stop])
+
+            # We can only match subsequences of the same size.
+            assert len(left) == len(right)
```
No new asserts pretty please!

Narsil replied:
Yes. I desperately need it while iterating as any breaking of this is hard to see, and breaks the algorithm very heavily. I'm still undecided wether to include that or not as a strong error. (it **should** not occur, but then the algorithm can still technically function without this but can be dramatically subpar...

sgugger replied:
You cna include an error, just with a test and a raise XxxError.

sgugger commented on tests/pipelines/test_pipelines_automatic_speech_recognition.py:
```diff
@@ -317,189 +316,123 @@ def test_return_timestamps_in_preprocess(self):
+        third_item = dataset["test"][3]
+        filename = third_item["file"]
+
+        # sp_audio = librosa.load(filename, sr=speech_recognizer.feature_extractor.sampling_rate)[0]
```
To clean up?

Narsil replied:
Yes.

sgugger reviewed (approved):
Looking good! I left a couple of comments, mainly to get more comments so we can follow the code :-)

Narsil commented:
Ok, I found that whisper doesn't play well **at all** without `return_timestamps:

```python
# Whisper HF captioning
from evaluate import load
from datasets import load_dataset
import numpy as np
from transformers import pipeline
import time
import whisper


libri = load_dataset("librispeech_asr", "clean", split="test")
model_name = "openai/whisper-large-v2"

speech_recognizer = pipeline(task="automatic-speech-recognition", model=model_name, framework="pt", device=2)

name = model_name.split("/")[-1][len("whisper") + 1 :]
model = whisper.load_model(f"{name}-v1" if name == "large" else name).to(device="cuda:3")

# Faulty example
start = 10
end = 15
iterator = range(start, end)
speech_file = np.concatenate([libri[i]["audio"]["array"] for i in iterator])
labels = " ".join(libri[i]["text"] for i in iterator)

for return_timestamps in [True, False]:
    print(f"========Timestamps ({return_timestamps})=========")
    start = time.time()
    hf = speech_recognizer(
        [speech_file],
        return_timestamps=return_timestamps,
        chunk_length_s=30,
        stride_length_s=[4, 4],
        batch_size=32,
        num_workers=1,
    )[0]
    end = time.time()
    print(f"model : {model_name}\nhf time : ", end - start)

    norm_labels = speech_recognizer.tokenizer._normalize(labels)
    norm_res = speech_recognizer.tokenizer._normalize(hf["text"])
    print("HF TEXT:", hf["text"])

    wer = load("wer")
    hf_wer = wer.compute(predictions=[norm_res], references=[norm_labels])
    print("hf wer :", hf_wer)

    start = time.time()
    openai = model.transcribe(np.asarray(speech_file, dtype=np.float32), without_timestamps=not return_timestamps)
    end = time.time()
    norm_open = speech_recognizer.tokenizer._normalize(openai["text"])
    print("openai time : ", end - start)
    openai_wer = wer.compute(predictions=[norm_open], references=[norm_labels])
    print("openai wer :", openai_wer)
    print("Openai TEXT:", openai["text"])

    # # if hf_wer > 1.5 * openai_wer:
    # #     import ipdb

    # #     ipdb.set_trace()
```

If you check, the output is really different in both cases. Whisper seems to stop outputting tokens when not using timestamps instead of continuing.

Narsil committed:
Some corrections.

Narsil committed:
Both fast and slow.

Narsil committed:
New black.

ArthurZucker commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -703,3 +703,291 @@ def get_decoder_prompt_ids(self, task=None, language=None, no_timestamps=True):
+                        chunks.append(chunk)
+
+                        # Flush all our temporary context
+                        # print("Flushing previous tokens (TIMESTAMP)")
```
```suggestion
```

ArthurZucker commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -703,3 +703,291 @@ def get_decoder_prompt_ids(self, task=None, language=None, no_timestamps=True):
+    all_special_ids = set(self.all_special_ids)
+    # - iterate over all outputs
+    for chunk_id, output in enumerate(model_outputs):
+        # We can drop everything to Python list, it's going to make
+        # our lives easier
```
```suggestion
```

ArthurZucker commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -703,3 +703,291 @@ def get_decoder_prompt_ids(self, task=None, language=None, no_timestamps=True):
+                # 3/ Timestamp token
+                time = (token - timestamp_begin) * time_precision + time_offset
+                time = round(time, 2)
+                if last_timestamp and token >= last_timestamp:
```
nit better naming for `last_timestamp` to make it  more understandable

ArthurZucker reviewed (commented):
This looks pretty good! I am curious about the latest benchmark results to see if this improved our accuracy or not? I like that this cleans up the ASR pipeline file, we could make it a little more general by having a `post_process_for_pipeline` method maybe WDYT? (instead of decode for asr) which would be called by `pipeline` all the time, and by default would do nothing (kind of like other utilities we have in tokenizer or generate)

The main function, `decode_asr`  seems a bit long, and I am not really  sure if processing lists is the best way (has of course pros and cons). 

I would like one of the tests to be kept as it was pretty reprensentative!

Narsil commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -703,3 +703,291 @@ def get_decoder_prompt_ids(self, task=None, language=None, no_timestamps=True):
+        )
+
+
+def _decode_asr(self, model_outputs, *, return_timestamps, return_language, time_precision):
```
I need to rename, I did that quickly to fix tests with the newly fast tokenizer :)

Narsil replied:
Done

Narsil commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -703,3 +703,291 @@ def get_decoder_prompt_ids(self, task=None, language=None, no_timestamps=True):
+                    if token >= timestamp_begin:
+                        # There can be several token in the right stride
+                        # But the last one is ALWAYS going to be skipped
+                        if (
+                            last_timestamp is not None
+                            and (token - timestamp_begin) * time_precision < right_stride_start
+                        ):
```
I am comparing 2 ints.

Well with numpy you are comparing the entire array, instead of just what you need. Performance is not important. And it would be harder to get exactly what you want here I think (since you need the last within stride OR last simply) since you cannot guarantee there's going to be a timestamp within stride) (There should, but there isn't necessarily)

Narsil committed:
Update src/transformers/models/whisper/tokenization_whisper.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

Narsil committed:
Update src/transformers/models/whisper/tokenization_whisper.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

Narsil commented on src/transformers/pipelines/automatic_speech_recognition.py:
```diff
@@ -82,112 +82,6 @@ def chunk_iter(inputs, feature_extractor, chunk_len, stride_left, stride_right,
             break
 
 
-def _find_timestamp_sequence(sequences, tokenizer, feature_extractor, max_source_positions):
```
It doesnt' handle, and cannot handle language though.

This is because this function returns a pure sequence of tokens, which still needs to be decoded specially by the tokenizer.

The new function handles properly the 4 arguments (return_timestamps:true/false, return_language:true/false) while this handles only 1 (and there is still complexity within `decode` which should be accounted for.

There's a reason I moved everything around, and as explicited in the PR, it's because I couldn't find any other way which could keep the current architecture. It's not about this particular piece of code.

ArthurZucker commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -703,3 +703,291 @@ def get_decoder_prompt_ids(self, task=None, language=None, no_timestamps=True):
+        # index = 0
+        max_ = 0.0
+        max_indices = (left_length, left_length, 0, 0)
+        # Here we're sliding matches
+        # [a, b, c, d]
+        #          [c, d, f]
+        # =        [c] == [d]
```
If you add `# docstyle-ignore` you can put everything under `"""` which is cleaner IMO than 30 `#` + we usually have these (it's a nit I guess)

ArthurZucker replied:
Maybe `[c] ?= [d]`, otherwise it's okay

sgugger reviewed (commented):
I have no problem with the refactor but I'm pretty concerned with the number of changes in the tests. This may mean this refactor is breaking things it shouldn't. Can we do the changes in the tests in a followup PR so that the diffs only show potential new tests, but no changes otherwise?

Narsil committed:
Remove print.

Narsil committed:
Undoing tests modifications.

Narsil committed:
Smaller test modifications.

Narsil committed:
Rename.

Narsil committed:
Remove maxDiff.

Narsil commented on tests/pipelines/test_pipelines_automatic_speech_recognition.py:
```diff
@@ -536,23 +538,23 @@ def test_whisper_timestamp_prediction(self):
                             "cut"
                         ),
-                        "timestamp": (5.5, 11.94),
+                        "timestamp": (5.5, 11.95),
```
@sgugger This should feel less scary ?

Also the audio sent here is 30s long, so the timestamp modifications looks like a fix (although given the unbounded nature of whisper, it's hard to be 100% sure about it it could very well be the model outputting weird stuff)

Narsil replied:
I'll wait for you to be OK with this before merging.

Narsil replied:
Ok I'll merge.

loretoparisi commented:
Folks I'm not sure where I'm wrong doing

```python
 predicted_ids = model.generate(input_features,
                                               max_length=model.config.max_target_positions,
                                               num_beams=5,
                                               length_penalty=1.0,
                                               repetition_penalty=1.2,
                                               do_sample=do_sample,
                                               temperature=temperature,
                                               return_timestamps=return_timestamps,
                                               early_stopping=early_stopping)
            
            if return_timestamps is False:
                # Decode the predicted IDs to text and append to results
                transcription_chunk = processor.batch_decode(predicted_ids,
                                                             task=task,
                                                             skip_special_tokens=skip_special_tokens,
                                                             return_timestamps=return_timestamps)[0]
                transcriptions.append(transcription_chunk)
            else:
                # Extract and decode the timing information
                timestamps = processor.batch_decode(predicted_ids,
                                                    task=task,
                                                    skip_special_tokens=skip_special_tokens,
                                                    return_timestamps=return_timestamps)[0]
```

Should I get `timestamps` information there when passing the kwargs  `return_timestamps` directly in the `batch_decode`?

loretoparisi commented:
My second attempt to retrieve timestamps

```python
             predicted_ids = model.generate(input_features,
                                               max_length=model.config.max_target_positions,
                                               num_beams=5,
                                               length_penalty=1.0,
                                               repetition_penalty=1.2,
                                               do_sample=do_sample,
                                               temperature=temperature,
                                               early_stopping=early_stopping)
            
               # Decode the outputs including the timestamps
                transcription_ids = predicted_ids[0].cpu().numpy().tolist()
                transcription_tokens = processor.batch_decode([transcription_ids])[0]

                segments = processor.tokenizer.convert_ids_to_tokens(transcription_ids)

                # Extract timing information from timestamp tokens
                transcription_chunk = []
                current_text = []
                current_start_time = None

                print(segments)
                for segment in segments:
                    if segment.startswith("<|startofprev|>"):
                        continue
                    if segment.startswith("<|endofprev|>"):
                        if current_text and current_start_time is not None:
                            end_time = (segments.index(segment) - 1) * HOP_LENGTH / sampling_rate
                            transcription_chunk.append({
                                "start_time": current_start_time,
                                "end_time": end_time,
                                "text": " ".join(current_text)
                            })
                            current_text = []
                            current_start_time = None
                    elif segment.startswith("<|") and segment.endswith("|>"):
                        continue
                    else:
                        if not current_text:
                            current_start_time = segments.index(segment) * HOP_LENGTH / sampling_rate
                        current_text.append(segment.replace("\u2581", " "))  # Replace special underscore token
```

This appraoch also failed, I cannot see in `segments` the timestamp tokens outputted.

loretoparisi commented:
THIRD ATTEMPT: Successful! I had to add `output_hidden_states`, `output_attentions` and the dicationary `return_dict_in_generate`

```python
                 predicted_ids = model.generate(input_features,
                                            max_length=model.config.max_target_positions,
                                            num_beams=5,
                                            length_penalty=1.0,
                                            repetition_penalty=1.2,
                                            do_sample=do_sample,
                                            temperature=temperature,
                                            return_dict_in_generate=True,
                                            output_scores=True,
                                            output_hidden_states=True,
                                            output_attentions=True,
                                            early_stopping=early_stopping)
            
                # Decode the outputs, including the timestamps
                decoded_transcription = processor.batch_decode(predicted_ids['sequences'], skip_special_tokens=False)[0]
                decoded_tokens = processor.tokenizer.tokenize(decoded_transcription)
                
                # Extract timestamps
                transcription_chunk = []
                current_text = []
                current_start_time = None
                for token in decoded_tokens:
                    if token.startswith("<|"):
                        if token not in {"<|startoftranscript|>", "<|transcribe|>", "<|notimestamps|>", "<|en|>", "<|endoftranscript|>"}:
                            if current_text:
                                end_time = start_idx / sampling_rate + processor.tokenizer.convert_tokens_to_ids([token])[0] * HOP_LENGTH / sampling_rate
                                transcription_chunk.append({
                                    "start_time": current_start_time,
                                    "end_time": end_time,
                                    "text": " ".join(current_text)
                                })
                                current_text = []
                            current_start_time = start_idx / sampling_rate + processor.tokenizer.convert_tokens_to_ids([token])[0] * HOP_LENGTH / sampling_rate
                    else:
                        if not current_text:
                            current_start_time = start_idx / sampling_rate
                        current_text.append(token.replace("Ġ", " "))  # Clean up token format

                if current_text:
                    end_time = len(audio_chunk) / sampling_rate
                    transcription_chunk.append({
                        "start_time": current_start_time,
                        "end_time": end_time,
                        "text": " ".join(current_text)
                    })
```
